src/lxc/usernsexec.c

```diff
@@ -0,0 +1,451 @@
+{
+	struct id_map *map;
+	long nsuid = 0, nsgid = 0;
+	int ret;
+
+	list_for_each_entry(map, active_map, head) {
+		if (map->idtype == ID_TYPE_UID) {
+			if (is_in_ns_range(nsuid, map))
+				nsuid += map->range;
+		} else {
+			if (is_in_ns_range(nsgid, map))
+				nsgid += map->range;
+		}
+	}
+
+	ret = add_map_entry(active_map, getgid(), nsgid, 1, ID_TYPE_GID);
+	if (ret < 0)
+		return false;
+	ret = add_map_entry(active_map, getuid(), nsuid, 1, ID_TYPE_UID);
+	if (ret < 0)
+		return false;
+	return true;
+}
+
+__hidden extern int lxc_log_fd;
+
+int lxc_usernsexec(int argc, char *argv[], int *status)
+{
+	int c, pid, ret;
+	char buf[1];
+	int pipe_fds1[2], /* child tells parent it has unshared */
+	    pipe_fds2[2]; /* parent tells child it is mapped and may proceed */
+	unsigned long flags = CLONE_NEWUSER | CLONE_NEWNS;
+	char ttyname0[256] = {0}, ttyname1[256] = {0}, ttyname2[256] = {0};
+	char *default_args[] = {"/bin/sh", NULL};
+	bool map_self = false;
+	LIST_HEAD(active_map);
+
+
+	lxc_log_fd = STDERR_FILENO;
+
+	if (isatty(STDIN_FILENO)) {
+		ret = readlink("/proc/self/fd/0", ttyname0, sizeof(ttyname0));
+		if (ret < 0) {
+			CMD_SYSERROR("Failed to open stdin");
+			return 1;
+		}
+
+		ret = readlink("/proc/self/fd/1", ttyname1, sizeof(ttyname1));
+		if (ret < 0) {
+			CMD_SYSINFO("Failed to open stdout. Continuing");
+			ttyname1[0] = '\0';
+		}
+
+		ret = readlink("/proc/self/fd/2", ttyname2, sizeof(ttyname2));
+		if (ret < 0) {
+			CMD_SYSINFO("Failed to open stderr. Continuing");
+			ttyname2[0] = '\0';
+		}
+	}
+
+	while ((c = getopt(argc, argv, "m:hs")) != EOF) {
+		switch (c) {
+		case 'm':
+			ret = parse_map(&active_map, optarg);
+			if (ret < 0)
+				return -1;
+			break;
+		case 'h':
+			return -1;
+		case 's':
+			map_self = true;
+			break;
+		default:
+			return -1;
+		}
+	};
+
+	if (list_empty(&active_map)) {
+		ret = find_default_map(&active_map);
+		if (ret < 0) {
+			fprintf(stderr, "Failed to find subuid or subgid allocation\n");
+			return 1;
+		}
+	}
+
+	// Do we want to support map-self with no other allocations?
+	// If so we should move this above the previous block.
+	if (map_self) {
+		if (!do_map_self(&active_map)) {
+			fprintf(stderr, "Failed mapping own uid\n");
+			return 1;
+		}
+	}
+
+	argv = &argv[optind];
+	argc = argc - optind;
+	if (argc < 1)
+		argv = default_args;
+
+	ret = pipe2(pipe_fds1, O_CLOEXEC);
+	if (ret < 0) {
+		CMD_SYSERROR("Failed to open new pipe");
+		return 1;
+	}
+
+	ret = pipe2(pipe_fds2, O_CLOEXEC);
+	if (ret < 0) {
+		CMD_SYSERROR("Failed to open new pipe");
+		close(pipe_fds1[0]);
+		close(pipe_fds1[1]);
+		return 1;
+	}
+
+	pid = fork();
+	if (pid < 0) {
+		CMD_SYSERROR("Failed to fork");
+		close(pipe_fds1[0]);
+		close(pipe_fds1[1]);
+		close(pipe_fds2[0]);
+		close(pipe_fds2[1]);
+		return 1;
+	}
+
+	if (pid == 0) {
+		close(pipe_fds1[0]);
+		close(pipe_fds2[1]);
+
+		opentty(ttyname0, STDIN_FILENO);
+		opentty(ttyname1, STDOUT_FILENO);
+		opentty(ttyname2, STDERR_FILENO);
```

Comment:
Same here. Why are we reopening the this? If there's no good reason we should kick it.